src/saml2/datetime/__init__.py

```diff
@@ -0,0 +1,202 @@
+"""Datetime structures and operations.
+
+This module encapsulates the structures that define datetime, time unit,
+duration and timezone objects, their relation and the operations that can be
+done upon them. Should these structures change all affected components should
+be under this module.
+
+There are three layers of specifications that define the structure and
+behaviour of time constructs for SAML2:
+
+- The SAML2-core specification - section 1.3.3 Time Values.
+  Reference: http://docs.oasis-open.org/security/saml/v2.0/saml-core-2.0-os.pdf
+
+- The W3C XML Schema Datatypes - section 3.2.7 dateTime.
+  Reference: https://www.w3.org/TR/xmlschema-2/#dateTime
+
+  [notable] W3C Date and Time Formats defines a profile of ISO 8601.
+  Reference: https://www.w3.org/TR/NOTE-datetime
+
+- The ISO 8601 standard upon which the dateTime datatype is based.
+  Reference: https://en.wikipedia.org/wiki/ISO_8601
+
+  [notable] Most systems implement rfc3339; a profile of ISO 8601.
+  Reference: https://tools.ietf.org/html/rfc3339
+
+Finally, further clarification was requested and in the following thread an
+answer was given by a member of the SAML Technical Committee:
+https://lists.oasis-open.org/archives/saml-dev/201310/msg00001.html
+
+To comply with the specifications, the existing implementations and the
+"unofficial errata" in the thread above, the following have been decided:
+
+- all ISO 8601 formats that can be parsed are accepted and converted to UTC.
+
+- if no timezone information is present, it is assumed that the other party is
+  following the current wording of the SAML2-core specification - the time is
+  assumed to be in UTC already, but "with no timezone component."
+
+- the datetime object produced are always in UTC timezone, that can be
+  represented as a string of ISO 8601 combined date and time format with
+  extended notation, where the timezone component is always present and
+  represented by the military timezone symbol 'Z'.
+"""
+
+import enum as _enum
+from datetime import datetime as _datetime
+
+from aniso8601 import parse_datetime as _datetime_parser
+
+import saml2.compat
+from saml2.datetime import duration
+from saml2.datetime import errors
+from saml2.datetime import timezone
+
+
+def parse(data):
+    """Return a datetime object in UTC timezone from the given data.
+
+    If timezone information is available the datetime object will be converted
+    to UTC timezone. If no timezone information is available, it will be
+    assumed to be in UTC timezone and that information will be added.
+
+    The parameter `data` is expected to be of type:
+    - datetime.datetime: a datetime.datetime object
+    - str: a string in ISO 8601 combined date and time format with extended
+      notation
+    - int: a number representing a POSIX timestamp
+    - float: a number representing a POSIX timestamp
+
+    The object returned is of type: datetime.datetime
```

Comment:
I like Python3 type hints, as they are inline and they refer to actual types - in contrast to types written as strings (wether in docstrings or [comments][cth]).

I was thinking about experimenting with [function annotations][fa] which should work with Python2 and still allow us to signal actual types rather than strings. It does have the drawback that we must type the parameter names, but I think it is still better.

  [cth]: https://mypy.readthedocs.io/en/latest/python2.html#type-checking-python-2-code
  [fa]: http://python-future.org/func_annotations.html

In the meantime, I can do the conversion to the format that the rest of the code follows.

Reply:
I would also prefer the Python 3 annotations and I haven't seen the function annotations that you link to. I wasn't thrilled about the syntax of those either but if you find it workable in your experiment I would be for them.